### DISCUSSION TITLE
How about officially supporting Solid.js?

### DISCUSSION THREAD

**Discussion by Souvlaki42** [shadcn-ui/ui#2262] [Ideas]
I was wandering how about adding support for Solid.js? Is the closest alternative to React that I can think of. I was trying to migrate my React app using Shadcn UI to Solid.js and the only real problem I faced is that the custom components I was previously using had some React specific features like HTMLAttributes and forwardRef. There is already an [unofficial port for Solid](https://www.solid-ui.com/) I just need something more official if possible.

Edit: React 19 and Next 15, deprecated forwardRef which could make the transition easier but it would still be nice for an official port to exist. Also this is one of the reasons as of why we need a framework agnostic way to build libraries nowadays.

**Comment by unomaas** [comment]
Bumping this, cause it's something I would also love to see! 🥳

**Comment by yanxi-tao** [comment]
+1

**Comment by totorofly** [comment]
+1

**Comment by BenjaminLindberg** [comment]
Very good idea! I'd also love to see ShadCN for SolidJS.

**Comment by whoiscarlo** [comment]
+1 Couldn't agree more!

**Comment by paxee** [comment]
This would be awesome!

**Comment by md-owes** [comment]
Any update from the maintainers of this repo?

**Comment by claughinghouse** [comment]
There are alternatives out there but the component parity is lacking. Really hopeful for official support!

https://www.solid-ui.com
https://shadcn-solid.com

**Comment by iitsg** [comment]
Would love this as well!

**Comment by squeaky-joe** [comment]
Reporting here for official SolidJS support. It'd be awesome and open lots of doors. While there are good alternatives they are still lacking.

**Comment by Ayothegod** [comment]
+1 
This package is really needed

**Comment by Ycg709jj** [comment]
+1

**Comment by dinesh-hills** [comment]
Would love to! +1

**Comment by leookun** [comment]
The concept of solid.js coincides with shadcn-ui. A natural couple. looking forward to official support！

**Comment by agustfricke** [comment]
Would love this! +1

**Comment by Borderliner** [comment]
This would be perfect! +1 SolidJs is awesome, please add support.

**Comment by koolkao** [comment]
+1

**Comment by mrinalwahal** [comment]
+1

**Comment by iskng** [comment]
deeply desired

**Comment by oddpxl** [comment]
Yep please pretty please

**Comment by 51522tt** [comment]
+1

**Comment by martincollado** [comment]
+1

**Comment by ivanjeremic** [comment]
Yes please.

**Comment by 0-don** [comment]
Would love this too, but the issue is that we need to choose a low-level UI components and primitives library for SolidJS.
Currently, I know of two options:

https://kobalte.dev/
https://corvu.dev/

If anyone knows of more, please share them here.

**Comment by EvvTim** [comment]
+1

**Comment by grimalschi** [comment]
More options available:
https://ark-ui.com
https://park-ui.com
https://github.com/hngngn/shadcn-solid (another unofficial port)
https://github.com/lxsmnsyc/terracotta
https://tanstack.com (table, form, virtual)
https://github.com/jundaoapp/design

**Reply by grimalschi** [reply]
Also Ariakit is being ported to Solid: https://github.com/ariakit/ariakit/issues/4117

**Reply by koolkao** [reply]
could some of these ports be incorporated back into the official Shadcn/UI?

**Reply by grimalschi** [reply]
It would be great, but it depends on @shadcn
The ideal solution, from my point of view, is to combine the efforts of https://github.com/hngngn/shadcn-solid and https://github.com/sek-consulting/solid-ui into an official port, same as Ariakit is doing now.

**Reply by Borderliner** [reply]
@grimalschi I hate the fact that Solid-ui depends heavily on Kobalte and Curvo. They're giant and make my app heavy.

**Reply by 0-don** [reply]
Yes, this is true, they make the app really big, which is a problem.

It would be nice if Radix would just make a SolidJS port https://github.com/radix-ui/primitives/discussions/1919. This would make the work for ShadCN much simpler.

**Reply by grimalschi** [reply]
@Borderliner shadcn-solid also relies on kobalte. But I have a feeling that primitives in Radix and Kobalte should weigh roughly the same?

**Reply by 0-don** [reply]
I get roughly the same build size as in react shadcnd as  with schadcn-solid which shouldnt be the case

**Reply by Borderliner** [reply]
@grimalschi Solidjs's philosophy is being lightweight, unlike react. I wouldn't use shadcn port of solidjs if it's heavy. I'd rather create my own components.

**Reply by grimalschi** [reply]
@Borderliner Being lightweight is cool, but the major thing in SolidJS is granular reactivity. I want to have it, but I also want to have a large set of ready to use components. So I think it doesn't goes against SolidJS principles.

**Comment by jarodtaylor** [comment]
Pretty please!!

**Comment by qingzhoufeihu** [comment]
deeply desired

**Comment by GATEHUA** [comment]
+1

**Comment by VEDA95** [comment]
+1

**Comment by SilentRhetoric** [comment]
+1

**Comment by mdhiego** [comment]
+1

**Comment by Souvlaki42** [comment]
I had no idea how many people wanted this library ported to Solid. We still don't have it though.

**Comment by rohancodex** [comment]
+1

**Comment by ShlokDesai33** [comment]
+1

**Comment by The-Best-Codes** [comment]
+1

**Comment by deviate-dv8** [comment]
+1, I hope 2025 is the year for this

**Comment by vadolasi** [comment]
+1

**Comment by MarvDann76** [comment]
+1
React is a bad fit for apps that have a parallel canvas rendering flow leveraging libraries such as fabric js.
SolidJS is a much better fit for apps like this.
Support for Solid would unlock some opportunities for us.

**Comment by taras-tereschenko** [comment]
+1

**Comment by raselldev** [comment]
i need something goodvibes like shadcn in solidjs

**Reply by The-Best-Codes** [reply]
In the meantime, use something like this:
https://www.solid-ui.com/

**Comment by sparklog** [comment]
+1

**Comment by shrubway** [comment]
+1

**Comment by penjj** [comment]
+1

**Comment by tmg0** [comment]
+1

**Comment by thanhnhahuynh1509** [comment]
+1

**Comment by DanteASC4** [comment]
+1

**Comment by dennypenta** [comment]
+42

**Reply by grimalschi** [reply]
Is this even legal? 😅

**Comment by Mohammed-Yasin-Mulla** [comment]
+1

**Comment by OmarShref** [comment]
+111111

**Comment by Souvlaki42** [comment]
It would be nice for Radix to be [official ported to Solid](https://github.com/radix-ui/primitives/discussions/1919), this would be make porting Shadcn to UI almost entirely automatic, due to how similar these libraries are,  but in the meantime, all these unofficial shadcn solid ports should merge into one and use a framework-agnostic headless UI library like [Ark UI](https://ark-ui.com/docs/overview/introduction).

**Comment by perry-autofi** [comment]
+1

**Comment by kurumeii** [comment]
+ 1

**Comment by badminimum** [comment]
+1

**Comment by BioHazard786** [comment]
+1

**Comment by nferreira1** [comment]
+1

**Comment by mixlette** [comment]
@Souvlaki42 @shadcn can we lock this conversation which is basically an endless flow of +1? Thank you.

**Reply by 0-don** [reply]
no +1

**Comment by ARMC0DE** [comment]
+1

**Comment by dmit4git** [comment]
+1

**Comment by SpeedoPasanen** [comment]
+1 lol 👍 
Clearly we need this.

**Comment by dennypenta** [comment]
The main problem is not in adopting shadcn, but porting headless part of radix UI. Looking at the experience of Kobalt UI maintainers it's not just adopting HTMLAttributes, there are lots more things.
If we manage to port Radix UI then support of Shadcn is not a question.
@Souvlaki42 Im happy to hear any ideas how to make it happen and what main challenges can be to port radix. At the moment I don't see it's viable.

**Comment by luoye00** [comment]
`https://github.com/hngngn/shadcn-solid/tree/main` may be can use.

**Comment by 0xrasla** [comment]
+1

**Comment by kuakman** [comment]
+1

**Comment by wangxg422** [comment]
+1. solidjs and shadcn/ui are both excellent frameworks. I hope they can be combined.

**Comment by thepooyan** [comment]
\+ 1

**Comment by theromanticsoul** [comment]
+1

**Comment by 0-don** [comment]
someone needs to fork [baseui](https://base-ui.com/) and keep it updated for solidjs

**Reply by taras-tereschenko** [reply]
yeah, i’ve been thinking about that too, even about using some ai code transformation tools. But we’d still need multiple people involved, the base ui repo just keeps getting bigger and there are more and more prs and issues coming in

**Reply by taras-tereschenko** [reply]
additionally, it must demonstrate a clear level of credibility and accountability for shadcn to consider including it as one of the base options offered to users

**Comment by msviderok** [comment]
For anyone interested in an unofficial (for now) port of Base UI for Solid: I’ve released the first iteration of what is intended to be a one-to-one, 100% DX-compatible port with v1.0.0-beta.1: [msviderok/base-ui-solid](https://github.com/msviderok/base-ui-solid)

There's also a [showcase demo](https://based-react-n-solid.vercel.app/) that demonstrates its usage for demos **and** layout for the whole website (if you switch the framework toggle). All I did was:
- created a project with [shadcn + Base UI starter](https://ui.shadcn.com/create?base=base) 
- changed `@base-ui/react` to `@msviderok/base-ui-solid` 
- changed `useState` to `createSignal` and passed values as `value()` instead of `value`

Interfaces for the components are supposed to be identical.

**Comment by carere** [comment]
Hello all 😄

I just released [Zaidan](https://zaidan.carere.dev) which is a port of ShadCn to SolidJS, it is based on Kobalte (which is like Radix for SolidJS), and I plan to add the base-ui port from @msviderok too (I already have the placeholder with a `coming soon` badge 😅).

I invite you all to try it out, I implemented the same customization that we have today on Shadcn "Create" page (base-color, theme-color, etc). Installation instructions are available for Astro, Tanstack Start, Solid Start & Vite.

Can't wait to have feedback 😃

**Reply by md-owes** [reply]
Why invest in creating another port using the same primitives as solid-ui? I'm thinking of adding a translation layer to the ShadCN repository that converts React wrappers into SolidJS wrappers. However, there may be another challenge with Radix primitives, as React wrappers use Radix and BaseUI under the hood, which are also written for React. If we can address these challenges, ShadCN could have official support for SolidJS. 

Let me know your thoughts.

**Reply by carere** [reply]
Both solid-ui & shadcn-solid are lacking updates since few month now. I started by translating components on the `dev` branch of Solid-ui repo then I decided to go with my own repo because I wanted full control over the implementation and release frequency. Also I plan on creating several more components (eg. AI Chat), based on updated primitive which are not available on either repos.
For the "translation layer", you can try, I translated all the primitive with an agent I created and optimized for doing this job. I plan on releasing this agent that way SolidJs community will be able to use it to translate their favorite shadcn based components.
IMO, if shadcn wants to officially support other framework, they'll start with Svelte & Vue since both framework are mainly supported by Vercel now.

**Reply by Borderliner** [reply]
Thank you, it's beautiful!

**Comment by sboagy** [comment]
+1!

**Comment by XhstormR** [comment]
use stenciljs to support all framework at time, https://github.com/stenciljs/core

**Comment by Prajwal-alphanimble** [comment]
you might wanna check out this discussion where I implemented a new resizable and dragable dashboard layout... this can be very helpfull to many developers in making analytic, sharable dashboard, the users can alter the dashboard based on their liking...

The disscussion link: https://github.com/shadcn-ui/ui/discussions/10466

**Comment by smirk-dev** [comment]
The timing of this discussion is really interesting now that Solid 2.0 has entered beta (released March 2026, https://github.com/solidjs/solid/discussions/2425). The 2.0 beta brings some meaningful changes to how reactivity and async work in Solid, and those changes actually make the case for an official shadcn port stronger, not weaker, because the community will need high-quality component primitives to build on top of a rewritten core.

That said, the practical reality is that maintaining an official port for a second framework would be a significant ongoing commitment for the shadcn team. There are already solid (pun intended) community options worth knowing about:

The most mature one right now is `shadcn-solid` (https://github.com/hngngn/shadcn-solid), which follows the same copy-paste philosophy as the original and is built on Kobalte, which is essentially the Radix UI equivalent for Solid. Kobalte handles accessibility primitives the same way Radix does, so the mental model transfers well.

There is also Solid UI (https://solid-ui.com), which ports both shadcn/ui and tremor-raw components using Kobalte and corvu.

The architectural blocker that keeps coming up in this thread, which is the React-specific APIs like `HTMLAttributes` and `forwardRef`, has actually gotten better over time. React 19 deprecated `forwardRef` in favour of passing `ref` as a regular prop, which nudges the component design toward something more framework-agnostic. It does not eliminate the gap, but it does reduce it.

If the shadcn team ever wanted to explore this direction, I think the most realistic path would be to first define a minimal framework-agnostic component contract (essentially: what props does each component expose, what are the accessibility semantics, what does the default styling look like), and then let community adapters implement that contract per framework. This is roughly what Base UI is doing with its headless primitives. The community ports for Solid are already converging on exactly this approach with Kobalte as the accessibility layer.

SolidStart 2.0-alpha.2 also shipped in February 2026, so the full-stack Solid ecosystem is maturing in lockstep. For anyone building new projects where framework choice is open, the Solid ecosystem deserves a much closer look than it was getting a year ago.

**Reply by zebbbra** [reply]
seems written with ai... we just discussed the fact that those repos are outdated above, like 4 messages up. 

[Zaidan](https://github.com/carere/zaidan) IS actually the most up to date repo.

**Reply by carere** [reply]
Omg, AI on github is really annoying...

**Comment by zebbbra** [comment]
[Zaidan](https://github.com/carere/zaidan) is actually the most up-to-date repo rn.

**Reply by Borderliner** [reply]
I hate building on top of Kobalt and Corvu. Just open up package.json for Zaidan and you'll see shit ton of dependencies there. I use SolidJs because it's lightweight. Now building a seemingly lightweight library on top of huge libraries don't look like lightweight to me. Don't believe me? Build an output and see the size.

**Reply by carere** [reply]
@Borderliner do you know how Shadcn works ? FYI, when installing components from Shadcn or Shadcn based registry (which Zaidan is), you copy paste the code, thus, the only dependencies installed are those required by the components. For example, if you install the button component, you'll install Kobalte, CVA, twmerge... That's not what I called lot of dependencies 😅

**Reply by md-owes** [reply]
I agree with @Borderliner. Initially, my product had over 450 components, and after introducing the Solid-UI framework, which internally uses Kobalte and Corvu, the total number of components on my development machine has increased to over 900—twice as many as I had before. I do not use lazy loading because I don't see much advantage in it. Instead, I prefer to load everything at once to reduce the number of API requests.

**Reply by carere** [reply]
@md-owes  Do you talk about dependencies or ui components ?
How can you double the amount of components by using Solid-Ui ?

**Reply by Borderliner** [reply]
> I agree with @Borderliner. Initially, my product had over 450 components, and after introducing the Solid-UI framework, which internally uses Kobalte and Corvu, the total number of components on my development machine has increased to over 900—twice as many as I had before. I do not use lazy loading because I don't see much advantage in it. Instead, I prefer to load everything at once to reduce the number of API requests.

Exactly the same experience. We had to ditch SolidUI in favor of hand-rolled components due to extreme size growth with Kobalte+Corvu. A simple pin-entry component increased the size of our bundle by a lot. After inspection, I realized the imported component internally imports tons of components that are needed to solve simple issues and are not tree-shaken.

**Reply by Borderliner** [reply]
> @Borderliner do you know how Shadcn works ? FYI, when installing components from Shadcn or Shadcn based registry (which Zaidan is), you copy paste the code, thus, the only dependencies installed are those required by the components. For example, if you install the button component, you'll install Kobalte, CVA, twmerge... That's not what I called lot of dependencies 😅

I don't use tanstack router in my stack, why should it be imported into my dependency tree? How can I opt out?

**Reply by sboagy** [reply]
Sorry if this is a dumb question.  But doesn't the whole app get tree-shaken for production?

**Reply by Borderliner** [reply]
> Sorry if this is a dumb question. But doesn't the whole app get tree-shaken for production?

It technically should. But you have to do a search recursively on components you use, and they end up importing tons of stuff anyway. My point is, libs like SolidUI and Zaidan provide easy-to-use conponents for you, but imo they're neither minimalist nor lightweight and not in line with SolidJs philosophy. If these don't matter to you, you'll be happy using them.

**Reply by md-owes** [reply]
@carere , I am talking about ui components when used in dev mode. I can literally see an additional 450 JavaScript files being requested by the browser to load all the Kobalte+Corvu components.
